File: app/src/main/java/com/wmods/wppenhacer/xposed/core/WaCallback.java
package com.wmods.wppenhacer.xposed.core;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class WaCallback implements Application.ActivityLifecycleCallbacks {
    private static void triggerActivityState(@NonNull Activity activity, WppCore.ActivityChangeState.ChangeType type) {
        WppCore.listenerAcitivity.forEach((listener) -> listener.onChange(activity, type));
    }

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle bundle) {
        WppCore.mCurrentActivity = activity;
        triggerActivityState(activity, WppCore.ActivityChangeState.ChangeType.CREATED);
        WppCore.activities.add(activity);
    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {
        WppCore.mCurrentActivity = activity;
        triggerActivityState(activity, WppCore.ActivityChangeState.ChangeType.STARTED);
        WppCore.activities.add(activity);
    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {
        WppCore.mCurrentActivity = activity;
        WppCore.activities.add(activity);
        triggerActivityState(activity, WppCore.ActivityChangeState.ChangeType.RESUMED);
    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {
        triggerActivityState(activity, WppCore.ActivityChangeState.ChangeType.PAUSED);
    }

    @Override
    public void onActivityStopped(@NonNull Activity activity) {
        triggerActivityState(activity, WppCore.ActivityChangeState.ChangeType.ENDED);
        WppCore.activities.remove(activity);
    }

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle bundle) {

    }

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {
        WppCore.activities.remove(activity);
    }
}

File: app/src/main/java/com/wmods/wppenhacer/xposed/core/WaCallback.kt
package com.wmods.wppenhacer.xposed.core

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import com.wmods.wppenhacer.xposed.core.ActivityStateRegistry.cleanup
import com.wmods.wppenhacer.xposed.core.ActivityStateRegistry.updateState
import com.wmods.wppenhacer.xposed.core.WppCore.ActivityChangeState
import com.wmods.wppenhacer.xposed.core.WppCore.listenerActivity
import java.util.function.Consumer

class WaCallback : ActivityLifecycleCallbacks {
    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
        WppCore.mCurrentActivity = activity
        updateState(activity, ActivityChangeState.ChangeType.CREATED)
        triggerActivityState(activity, ActivityChangeState.ChangeType.CREATED)
    }

    override fun onActivityStarted(activity: Activity) {
        WppCore.mCurrentActivity = activity
        updateState(activity, ActivityChangeState.ChangeType.STARTED)
        triggerActivityState(activity, ActivityChangeState.ChangeType.STARTED)
    }

    override fun onActivityResumed(activity: Activity) {
        WppCore.mCurrentActivity = activity
        updateState(activity, ActivityChangeState.ChangeType.RESUMED)
        triggerActivityState(activity, ActivityChangeState.ChangeType.RESUMED)
    }

    override fun onActivityPaused(activity: Activity) {
        updateState(activity, ActivityChangeState.ChangeType.PAUSED)
        triggerActivityState(activity, ActivityChangeState.ChangeType.PAUSED)
    }

    override fun onActivityStopped(activity: Activity) {
        updateState(activity, ActivityChangeState.ChangeType.ENDED)
        triggerActivityState(activity, ActivityChangeState.ChangeType.ENDED)
    }

    override fun onActivityDestroyed(activity: Activity) {
        updateState(activity, ActivityChangeState.ChangeType.DESTROYED)
        triggerActivityState(activity, ActivityChangeState.ChangeType.DESTROYED)
        cleanup()
    }

    override fun onActivitySaveInstanceState(activity: Activity, bundle: Bundle) {
    }

    companion object {
        private fun triggerActivityState(activity: Activity, type: ActivityChangeState.ChangeType) {
            listenerActivity.forEach(Consumer { listener: ActivityChangeState? ->
                listener!!.onChange(
                    activity,
                    type
                )
            })
        }
    }
}
